[quant] Dynamic quantization for bias.

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26057 [quant] Dynamic quantization for bias.**

bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

![% Change (With vs Without) (-ve change means faster)](https://user-images.githubusercontent.com/37562707/64818243-6d5e8b00-d560-11e9-9158-778cdda4f610.png)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

dskhudia committed:
[quant] Dynamic quantization for bias.

bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

TODO: Add some performance numbers.

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

jianyuh commented:
Saw the OSS failure:
```
Sep 11 22:19:16 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:195:39: error: no matching function for call to 'fbgemm::ReQuantizeOutput<true, (fbgemm::QuantizationGranularity)2, float, int, fbgemm::DoNothing<float, float> >::ReQuantizeOutput(fbgemm::DoNothing<unsigned char, unsigned char>&, float*, int32_t&, int32_t&, int*, int32_t*, int*, const float*&, int64_t&, int, float*)'
Sep 11 22:19:16                act_times_w_scale.data());
Sep 11 22:19:16                                        ^
```
Is it because the FBGEMM code changes are not propagated to PyTorch?

dskhudia commented:
> Saw the OSS failure:
> 
> ```
> Sep 11 22:19:16 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:195:39: error: no matching function for call to 'fbgemm::ReQuantizeOutput<true, (fbgemm::QuantizationGranularity)2, float, int, fbgemm::DoNothing<float, float> >::ReQuantizeOutput(fbgemm::DoNothing<unsigned char, unsigned char>&, float*, int32_t&, int32_t&, int*, int32_t*, int*, const float*&, int64_t&, int, float*)'
> Sep 11 22:19:16                act_times_w_scale.data());
> Sep 11 22:19:16                                        ^
> ```
> 
> Is it because the FBGEMM code changes are not propagated to PyTorch?

Yes. Once the fbgemm diffs land, it should work fine.

dskhudia committed:
Update on "[quant] Dynamic quantization for bias."


bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

TODO: Add some performance numbers.

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

jianyuh commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -165,16 +142,22 @@ class QLinearInt8 final : public torch::OperatorKernel {
+          fbgemm::QuantizationGranularity::TENSOR,
+          float>
+          outputProcObj(
+              doNothingObj,
```
Maybe keep the comments for better annotating these arguments.

dskhudia replied:
I removed the trivial ones and kept ones which provide some context.

jianyuh commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -196,17 +179,20 @@ class QLinearInt8 final : public torch::OperatorKernel {
-              /*col_offsets=*/col_offsets.data(),
-              /*bias=*/bias_ptr,
-              /*nCol=*/N);
+              doNothingObj,
```
Ditto.

dskhudia replied:
Same reply as above.

jianyuh commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -229,16 +229,9 @@ def from_float(cls, mod):
         wt_scale, wt_zp = weight_observer.calculate_qparams()
-        # Scale bias to activation_scale/2^16, this quantizes bias
-        # to about 24 bits of precision
-        bias_scale = float(act_scale / (2**16))
```
The bias handler logic in the new FBGEMM API is different from "activation_scale/2^16" here. Will this cause any numerical diffenrece?

jianyuh replied:
Just saw your comment that the bias is just not quantized, so I guess the accuracy should be improved with the new API.

raghuramank100 replied:
This was a temporary change, any numeric change will be below 1 LSB as bias was previously quantized to ~24 bit precision.

raghuramank100 commented:
Looks good, have you also checked the output of the  test: caffe2/test/test_quantized_models.py to ensure that overall numerics at the model level stay close?

raghuramank100 reviewed (approved):
Looks good, please check end to end model numerics before submitting.

dskhudia commented:
> Looks good, have you also checked the output of the test: caffe2/test/test_quantized_models.py to ensure that overall numerics at the model level stay close?

Yes. That test runs as a part of test:quantized and the results for this are shown in the description above.

jianyuh reviewed (approved):
LGTM!

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -133,53 +133,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
-      auto bias_contig = qbias.contiguous();
-      bias_ptr =
-          reinterpret_cast<int32_t*>(bias_contig.data_ptr<c10::qint32>());
+      bias_ptr = reinterpret_cast<float*>(bias.data_ptr<float>());
```
you don't need reinterpret_cast if it's float

dzhulgakov commented on torch/nn/quantized/functional.py:
```diff
@@ -77,7 +77,7 @@ def conv2d(input, weight, bias,
         input: quantized input tensor of shape :math:`(\text{minibatch} , \text{in\_channels} , iH , iW)`
         weight: quantized filters of shape :math:`(\text{out\_channels} , \frac{\text{in\_channels}}{\text{groups}} , kH , kW)`
-        bias: **non-quantized** bias tensor of shape :math:`(\text{out\_channels})`. The tensor type must be `torch.int32`.
+        bias: **non-quantized** bias tensor of shape :math:`(\text{out\_channels})`. The tensor type must be `torch.float`.
```
@supriyar - is qnnpack also expecting fp32 bias? I think it's best to keep them the same. (fp32 itself sounds fine, we can re-quantize inside pack)

If they are different, given that quantization of bias is cheap, we could allow this function to receive either quantized or non-quantized bias and do conversion internally if needed.

supriyar replied:
qnnpack expects int32 bias. We prepack weight and bias together in qnnpack 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/pack.h#L81

We decided to not change qnnpack (in the interest of time) to decouple this and instead pack bias in the fbgemm struct to make them semantically the same.
Given that I think 2nd option proposed by @dzhulgakov makes sense - i.e. change the function to accept both quantized and non-quantized bias.

dzhulgakov reviewed (commented):
@supriyar - Let's make sure that it's qnnpack-compatible

dskhudia committed:
Update on "[quant] Dynamic quantization for bias."


bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

![% Change (With vs Without) (-ve change means faster)](https://user-images.githubusercontent.com/37562707/64818243-6d5e8b00-d560-11e9-9158-778cdda4f610.png)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

dskhudia committed:
Update on "[quant] Dynamic quantization for bias."


bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

![% Change (With vs Without) (-ve change means faster)](https://user-images.githubusercontent.com/37562707/64818243-6d5e8b00-d560-11e9-9158-778cdda4f610.png)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -48,7 +48,7 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
     TORCH_CHECK(
         weight.dim() == 2,
-        "The weight tensor for quantized::linear_prepack should"
+        "The weight tensor for quantized::linear_prepack (fbgemm) should"
```
revert? - it's not a specific requirement for fbgemm :)

dskhudia replied:
Not a part of this diff now.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -133,53 +133,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
-      auto bias_contig = qbias.contiguous();
-      bias_ptr =
-          reinterpret_cast<int32_t*>(bias_contig.data_ptr<c10::qint32>());
+      bias_ptr = bias.data_ptr<float>();
```
I'd add a nice TORCH_CHECK that the input is unquantized with a human-understandable message. data_ptr would do one for you but it's nicer to call out that we expect bias in fp32

dskhudia replied:
Added a check.

dskhudia committed:
Update on "[quant] Dynamic quantization for bias."


bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

![% Change (With vs Without) (-ve change means faster)](https://user-images.githubusercontent.com/37562707/64818243-6d5e8b00-d560-11e9-9158-778cdda4f610.png)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

[ghstack-poisoned]

dskhudia committed:
Update on "[quant] Dynamic quantization for bias."


bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

![% Change (With vs Without) (-ve change means faster)](https://user-images.githubusercontent.com/37562707/64818243-6d5e8b00-d560-11e9-9158-778cdda4f610.png)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

[ghstack-poisoned]
